apps/src/p5lab/spritelab/commands/validationCommands.js

```diff
@@ -57,10 +65,722 @@ export const commands = {
+      this.previous.sprites.push({
+        id: spriteId,
+        costume: this.nativeSpriteMap[spriteId].getAnimationLabel(),
+        x: this.nativeSpriteMap[spriteId].x,
+        y: this.nativeSpriteMap[spriteId].y,
+        behaviors: this.getBehaviorsForSpriteId(spriteId),
+        tint: this.nativeSpriteMap[spriteId].tint || '',
+        scale: this.nativeSpriteMap[spriteId].getScale(),
+        speed: this.nativeSpriteMap[spriteId].speed,
+        rotation: this.nativeSpriteMap[spriteId].rotation
+      });
+    }
+  },
+
+  // If the student has not completed any criteria, they are "failing".
+  // This determines things like the current color of the validation timer bar below the play area.
+  getPassState(criteria) {
+    var state = 'pass';
+    for (const criterion in criteria) {
+      if (!criteria[criterion].complete) {
+        state = 'fail';
+      }
+    }
+    return state;
+  },
+
+  calculateBarScale(state) {
+    let scale = APP_WIDTH;
+    switch (state) {
+      case 'pass':
+        scale =
+          APP_WIDTH /
+          (this.validationFrames.pass - this.validationFrames.successFrame);
+        break;
+      case 'fail':
+        scale = APP_WIDTH / this.validationFrames.fail;
+        break;
+      default:
+        scale = APP_WIDTH;
+        break;
+    }
+    return scale;
+  },
+
+  // checkAllCriteria() is used in levels, typically occuring once per frame
+  // as part of updateValidation(). Validation is essentially an ordered
+  // list of one or more "criteria", each consisting of a predicate function and
+  // a feedback message key. The predicate functions run each tick, and if true,
+  // the criteria is marked as completed. At the end of the program, if all the
+  // criteria are complete, the student passes the level. If any criteria are
+  // not complete, the student sees the feedback message corresponding to the
+  // first unmet criteria.
+  checkAllCriteria(criteria) {
+    for (const criterion in criteria) {
+      if (!criteria[criterion].complete) {
+        if (criteria[criterion].predicate()) {
+          criteria[criterion].complete = true;
+        }
+      }
+    }
+  },
+
+  // Used at the end of a level. Find the first failed criteria and pass associated feedback.
+  reportFailure(criteria) {
+    let firstFailed = -1;
+    for (const criterion in criteria) {
+      if (!criteria[criterion].complete && firstFailed === -1) {
+        firstFailed = criterion;
+      }
+    }
+    if (firstFailed > -1) {
+      return criteria[firstFailed].feedback;
+    }
+  },
+
+  // Used at the end of a level. If there are no failed criteria, pass the generic success feedback.
+  reportSuccess(criteria) {
+    let firstFailed = -1;
+    for (const criterion in criteria) {
+      if (!criteria[criterion].complete && firstFailed === -1) {
+        firstFailed = criterion;
+      }
+    }
+    if (firstFailed === -1) {
+      return 'genericSuccess';
+    }
+  },
+
+  // CRITERIA FUNCTIONS
```

Comment:
Sorry for not thinking of this earlier: would it make sense to separate these functions into a separate file?  They have a slightly different scope of responsibility to the above functionality, and this file is quite large already.

Reply:
Also a note that I didn't focus so strongly on these in my code review, but they looked good from a fast scan.

Reply:
Good idea! Moved to a separate file: https://github.com/code-dot-org/code-dot-org/blob/8b3cc7410c08f3a5defa1bd3b87ed96a807476d2/apps/src/p5lab/spritelab/commands/criterionCommands.js

I would imagine that this new file would be the one that needs additions later, to make sure the curriculum team is unblocked from validating new lessons. In other words, the "criteria" paradigm wouldn't change much, but we might add more functions to check for different things that aren't part of this PR.